keps/sig-storage/20190530-pv-health-monitor.md

```diff
@@ -0,0 +1,577 @@
+* Volume could be out of capacity.
+* The disk may be degrading which affects its performance.
+
+If the volume is mounted on a pod and used by an application, the following problems could also happen.
+* There may be read/write I/O errors.
+* The file system on the volume may be corrupted.
+* Filesystem may be out of capacity.
+* Volume may be unmounted by accident outside of Kubernetes.
+
+If the CSI driver has implemented the CSI volume health function proposed in this design document, Kubernetes could communicate with the CSI driver to retrieve any errors detected by the underlying storage system. Kubernetes can store this data on the associated PVC so that user can inquire this information and decide how to handle them. For example, if the volume is out of capacity, user can request a volume expansion to get more space. Kubernetes may also use volume health information stored on the PVC to automatically reconcile.
+
+There could be conditions that cannot be reported by a CSI driver. There could be network failure where Kubernetes may not be able to get response from the CSI driver. In this case, a call to the CSI driver may time out. There could be network congestion which causes slow response. One or more nodes where the volume is attached to may be down. This can be monitored and detected by the volume health controller so that user knows what has happened.
+
+The Kubernetes components that may modify the PVC status with volume health information include the following:
+
+* An external monitoring agent on each kubernetes worker node.
+* An external monitoring controller on the master node.
+
+Details will be described in the following proposal section.
+
+## Proposal
+
+Volume monitoring is the main focus of this proposal. Reactions are not in the scope of this proposal.
+
+- If a volume provisioned by the CSI driver is deleted, we need to mark the corresponding PVC to inform users.
+
+- If mounting error occurs, we need to mark the PVC.
+
+- If other errors occur, we also need to mark the PVC.
+
+The main architecture is shown below:
+
+![pv health monitor architecture](./pv-health-monitor.png)
+
+The following areas will be the focus of this proposal at first:
+
+- The health condition checking of volumes themselves. For example, whether the volume is deleted, whether the usage is reaching the threshold, and so on.
+- Mounting conditions checking.
+- Other errors that could affect the usability of the volume.
+
+Three main parts are involved here in the architecture.
+
+- API changes: The Status field in the PVC will be used to mark volumes if they are unhealthy.
+
+- External Controller:
+  - The external controller will be deployed as a sidecar together with the CSI controller driver, similar to how the external-provisioner sidecar is deployed.
+  - Trigger controller RPC to check the health condition of network volumes themselves for network storage.
+  - Watch for node failure events for both network and local storage.
```

Comment:
Is this part going to be part of the sidecar too? Or is that an independent controller?

Reply:
So right now the idea is to just do this check inside the external controller sidecar.